src/Data/Version/Compat.hs

```diff
@@ -0,0 +1,52 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module Data.Version.Compat (
+  module Base
+, makeVersion
+) where
+import Data.Version as Base
+
+#if !MIN_VERSION_base(4,7,0)
+import Data.Data
+#endif
+
+#if !MIN_VERSION_base(4,8,0)
+import Prelude.Compat
+
+# if __GLASGOW_HASKELL__ >= 708
+import GHC.Exts (IsList(..))
+# endif
+#endif
+
+#if !MIN_VERSION_base(4,7,0)
+versionConstr :: Constr
+versionConstr = mkConstr versionDataType "Version" ["versionBranch","versionTags"] Prefix
+
+versionDataType :: DataType
+versionDataType = mkDataType "Data.Version.Version" [versionConstr]
+
+instance Data Version where
+  gfoldl k z (Version bs ts) = z Version `k` bs `k` ts
+  toConstr (Version _ _) = versionConstr
+  gunfold k z c = case constrIndex c of
+                    1 -> k (k (z Version))
+                    _ -> error "Data.Data.gunfold(Version)"
+  dataTypeOf _  = versionDataType
```

Comment:
No idea why it's hand-written. All I can find is [this thread](http://permalink.gmane.org/gmane.comp.lang.haskell.libraries/20269) from a mailing list.